Lombiq.Tests.UI/Services/UITestExecutionSession.cs

```diff
@@ -481,7 +481,7 @@ private async Task SetupAsync()
 
             _dockerConfiguration = TestConfigurationManager.GetConfiguration<DockerConfiguration>();
 
-            var resultUri = await _currentSetupSnapshotManager.RunOperationAndSnapshotIfNewAsync(async () =>
```

Comment:
Please keep this. The setup operation needs to be able to create the URL the test should start from.

Reply:
Okay, and in what form is it used currently? How can UI tests be started on different URLs currently? I can't seem to find it.

Reply:
It's this feature you removed here. The URL returned by the setup operation will be the starting URL of the test.

Reply:
I luled, obviously this is what it was using it, but it is not anymore. So let me rephrase my question. Why is it needed to use this URL to navigate to the relative URL at the end of the setup? Is there a reason for that? If so, I don't think `resultUri.PathAndQuery` is the right way of doing it.

So e.g. what you say is if you set a URL of https://example.com/path/ to start the test on, you would get the relative Url `https://example.com/path/path` at line `await _context.GoToRelativeUrlAsync(resultUri.PathAndQuery);` which is obviously not okay.
So the correct navigation step is to navigate to `/`, which means `resultUri` isn't needed.

Reply:
Why would it be correct to always navigate to `/`? The reason we have the setup operation return a URL is precisely because that's not necessarily the case: the URL it returns is the start URL of the test. This is the URL that'll be open when control arrives at the beginning of a test's code.  The setup operation is not necessarily just filling out the setup screen, it can be anything to initialize the app to a testable state.

Reply:
Did you understand my example?

Reply:
> the URL it returns is the start URL of the test. This is the URL that'll be open when control arrives at the beginning of a test's code. The setup operation is not necessarily just filling out the setup screen, it can be anything to initialize the app to a testable state.

Where do you set this URL currently?

Reply:
This URL is the URL to start the test on. The app may launch with a different URL, i.e. the _setup_ may start with one URL, but the _test_ can start with a second one. That's what the setup operation returns.

In any case, feel free to change whatever is necessary to make the prefixed base path work, but do not remove the ability for the setup operation to return a custom start URL for a test.

Reply:
Could you show me an example where this is done? Because I don't understand this scenario.

Reply:
It's in every setup operation, since otherwise you can't compile, e.g.: https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI.Samples/Helpers/SetupHelpers.cs#L46

Reply:
Okay, now I understand what you mean, but it is really weird to do this:
![image](https://github.com/user-attachments/assets/a70f8012-ce68-48ef-9f0b-e5a201b01371)
Let's say you set the basepath to start the UI tests on for `/cms` and to operate with this basepath for everything. It means the setup will be started on that page and it will return `https://localhost:port/cms/` as the Uri (homePageUri variable here), and if I return that because I want to be on that page after the setup process then it will redirect to `https://localhost:port/cms/cms`. 

So I have to return a `https://localhost:port/` Uri, because in [GoToRelativeUrlAsync](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L25), the [GetAbsoluteUri](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L71) is called to create the Uri based on the `context.Scope.BaseUri` which is set to `https://localhost:port/cms/`. So I return  `https://localhost:port/` to get `https://localhost:port/cms` at `UITestExecutionSession` `await _context.GoToRelativeUrlAsync(resultUri.PathAndQuery);`

Or is this only weird for me? I can sleep with this in my mind, it's just weird.

Reply:
You can try `_context.GoToRelativeUrlAsync("/" + resultUri.PathAndQuery);` if `PathAndQuery` never contains a leading slash. That's also semantically more appropriate, since the assumption there is that the URL returned from the setup operation is always relative to the app root.

Reply:
Yeah, that was my original thought. But that won't work because [GetAbsoluteUri](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs#L71) does `.TrimStart('/')` to the given Uri.

Reply:
I discovered that we actually already have a mechanism for all of this, called `UrlPrefix`. So, rather do what [`SwitchCurrentTenant`](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Services/UITestContext.cs#L246) does. Everything else should work without changes.

Reply:
I swear I did not spend several hours on this.

Reply:
Uhm this actually won't work because we also need a different base path during setup.

Reply:
Okay nvm, I can still do it with this, but now it just seems like another solution, and not necessarily a better solution. But I can do it this way if you say so.

Reply:
Utilizing an existing mechanism, just also earlier, seems like the solution to go, yes.